rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,226 @@
+s3://<table_storage_bucket>/0bfb3d6e/<hudi_table_name>/.075f3295-def8-4a42-a927-07fd2dd2976c-0_7-11-49_20220301005056692.log.1_0-22-26
+...
+```
+**Note**: For `Merge on Read` tables, the log files will also go through the same hashing process and may not end up under
+the same prefix as the base parquet file for the FileSlice to which it belongs.
+
+The original table path will continue to store the `metadata folder` and `partition metadata` files:
+```
+s3://<table_bucket>/<hudi_table_name>/.hoodie/...
+s3://<table_bucket>/<hudi_table_name>/country=usa/.hoodie_partition_metadata
+s3://<table_bucket>/<hudi_table_name>/country=india/.hoodie_partition_metadata
+s3://<table_bucket>/<hudi_table_name>/country=uk/.hoodie_partition_metadata
+...
+```
+
+#### Hashing
+
+#####Option 1:
+We can re-use the implementations is `HashID` class to generate hash on `File Name` or `Partition + File Name`, which
+uses XX hash function with 32/64 bits (known for being fast).
+
+#####Option 2:
+To generate the prefixes we can use `Murmur 32 bit` hash, which is known for being fast and provides good distribution
+guarantees. We might have to further do bucketing and re-hash it to reduce the number of possible hashes from 2^32 to a
+slightly lower number, as it may be overkill to have that many unique hashes, which might result in scenarios
+where each file is under a different prefix.
+
+The hashing function should be made user configurable.
+
+### Maintain mapping to files
+
+In [RFC-15](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=147427331), we introduced an internal
+Metadata Table with a `files` partition that maintains mapping from partitions to list of files in the partition stored
+under `Table Path`. This mapping is kept up to date, as operations are performed on the original table. We will leverage
+the same to now maintain mappings to files stored at `Table Storage Path` under different prefixes.
+
+Here are some of the design considerations:
+
+1. Metadata table is a pre-requisite for federated storage to work. Since Hudi 0.11 we have enabled metadata table by
+default and hence this feature can be enabled by the users as long as they are not explicitly turning off metadata
+table, in which case we should throw an exception.
+
+2. The federated storage cannot be enabled on an existing table that is already bootstrapped with Hive storage
+layout. To switch to federated storage, the table will need to be re-bootstrapped with the new layout.
+
+3. The Instant metadata (`HoodieCommitMetadata`,`HoodieCleanMetadata` etc.) will always act as the source of file listing
+for metadata table to be populated.
+
+4. `HoodieCommitMetadata` currently stores `file name` instead of complete `file path`. We will have to modify commit
+metadata to store the complete file path instead of just file name, as the files are now distributed across several random
+prefix paths instead of a derivable table/partition path.
```

Comment:
Seems not necessary ? The new prefix works just like another `partition` path,  and the probability of conflict is same as the normal partitioned table, we can calculate the hash prefix on the fly when fetching these files.

Reply:
+1 

Storing absolute paths is a can of worms we want to avoid opening at all costs.

Before diving into why it should be avoided, i want to step back and first get alignment on the following critical premise -- any federation should be _completely transparent_ to all of the Hudi components, ie all of the components (beside File Access layer) should not be even aware that there's any storage federation happening: we're essentially should create a _virtual FS view_ of the table where all files are still co-located w/in the same partition, and _real FS view_ where they could be stored following arbitrary strategy. That way we can guarantee, that no high-level component relying on provided FS views will be impacts and will still function as expected.

Now with that in mind, i envision following issues if we depart from this tenet:

 - Biggest issues with absolute paths is that storing absolute paths exposes other Hudi components to the table layout and requires them to be aware of how the table is stored which will quickly go out of hand as the number of strategies will grow. We should make sure that higher-level components are not exposed to the storage layout and have their own view of the table which is still having files w/in the partition co-located (see paragraph above for more details)

 - Storing absolute paths will also make table's relocations impossible (it's not out of the realm of possibilities that you might want to relocate your table from one bucket to another, to switch regions for ex)

 - It will also entail increased storage footprint (in MT for ex)

Reply:
+1 at least for the hashing prefix-based approach, I don't see it's necessary to store absolute paths.